### PR TITLE
EKF: support passing in an external position estimate

### PR DESCRIPTION
**Pull request by tridge**
This allows for a position update from an external source such as someone spotting a landmark with an FPV camera. 
MAVProxy support here:
https://github.com/ArduPilot/MAVProxy/pull/1186
needs this mavlink PR: https://github.com/ArduPilot/mavlink/pull/318

### CODE REVIEW

File: libraries/GCS_MAVLink/GCS_Common.cpp [hunk 1]
```diff
@@ -4980,6 +4981,34 @@ MAV_RESULT GCS_MAVLINK::handle_command_int_do_set_home(const mavlink_command_int
...
+    if (!location_from_command_t(p2, loc)) {
+        return MAV_RESULT_DENIED;
+    }
+    uint32_t timestamp_ms = correct_offboard_timestamp_usec_to_ms(uint64_t(p2.param1*1e6), PAYLOAD_SIZE(chan, COMMAND_INT));
+    const uint32_t processing_ms = p2.param2*1e3;
+    const float pos_accuracy = p2.param3;
+    if (timestamp_ms > processing_ms) {
+        timestamp_ms -= processing_ms;
+    }
+    if (!AP::ahrs().handle_external_position_estimate(loc, pos_accuracy, timestamp_ms)) {
```
Comment (priseborough) [inline libraries/GCS_MAVLink/GCS_Common.cpp:5006]: The pos_accuracy and timestamp_ms are being passed in reverse order

**Comment by priseborough** [comment]
See https://github.com/tridge/ardupilot/pull/41

File: libraries/GCS_MAVLink/GCS_Common.cpp (hunk 1 above)
Comment (tridge) [reply libraries/GCS_MAVLink/GCS_Common.cpp:5006]: oops! fixed

**Comment by rmackay9** [comment]
Very interesting, I was thinking about doing something like this with an image tracking camera.

Why wouldn't we go through the AP_VisualOdometry library though?

**Comment by tridge** [comment]
> Why wouldn't we go through the AP_VisualOdometry library though and use one of the many existing vision-position messages?

a few reasons. First, we want the ACK/NACK as it is nacked when we are not in AID_NONE. Secondly, we need the dual timestamp to cope with much longer than normal lag - it would be beyond fusion horizon if done via vision msgs.
Finally, it is a reset, not a normal fusion, so it leaves velocity alone. A vision position will change velocity.

**Comment by tridge** [comment]
log from a CubeOrange quadplane (MakeFlyEasy Hero):
http://uav.tridgell.net/tmp/00000068.BIN

**Comment by priseborough** [comment]
The positon and velocity states and their corresponding variances from the quadplane test flight have the expected behaviour, eg:

Postion resets do not directly affect the velocity states

Screen Shot 2023-05-28 at 5 53 57 pm
Screen Shot 2023-05-28 at 5 54 18 pm

The position state variance reset to the expected value as specified by p3 in the mavlink command
Screen Shot 2023-05-28 at 5 54 24 pm

Velocity state variances do not change
Screen Shot 2023-05-28 at 5 55 17 pm

**Review by priseborough — APPROVED** [review]

**Comment by rmackay9** [comment]
I'd like to discuss this a bit before we merge.  Obviously the functionality is great, it's just the architecture and possibly message name.  Also I've got a similar change that adds a position reset for vision-position sources and I'd like us to make sure they're consistent.  PR coming for that later today.

**Comment by rmackay9** [comment]
Here's my related PR:

1. mavlink PR https://github.com/ArduPilot/mavlink/pull/319 adds a new align-vision-position-local-ned command
2. flight code PR https://github.com/ArduPilot/ardupilot/pull/23911 supports correcting the drift from a vision position system

File: libraries/GCS_MAVLink/GCS_Common.cpp [hunk 2]
```diff
@@ -4980,6 +4981,34 @@ MAV_RESULT GCS_MAVLINK::handle_command_int_do_set_home(const mavlink_command_int
     return MAV_RESULT_ACCEPTED;
 }
 
+MAV_RESULT GCS_MAVLINK::handle_command_int_external_position_estimate(const mavlink_command_int_t &packet)
+{
+
+    if (!isnan(packet.z)) {
```
Comment (tridge) [inline libraries/GCS_MAVLink/GCS_Common.cpp:4987]: check frame

**Comment by tridge** [comment]
needs #if for flash cost, 2M boards only

File: libraries/GCS_MAVLink/GCS_Common.cpp [hunk 3]
```diff
@@ -4980,6 +4981,34 @@ MAV_RESULT GCS_MAVLINK::handle_command_int_do_set_home(const mavlink_command_int
...
+        // ensure sender doesn't want altitude, which we don't support
+        return MAV_RESULT_UNSUPPORTED;
```
Comment (sypaq-nexton) [inline libraries/GCS_MAVLink/GCS_Common.cpp:4989]:
```suggestion
        return MAV_RESULT_DENIED;
```
It is supported, it just has invalid params.

File: libraries/GCS_MAVLink/GCS_Common.cpp (hunk 2 above)
Comment (tridge) [reply libraries/GCS_MAVLink/GCS_Common.cpp]: done

File: libraries/GCS_MAVLink/GCS_Common.cpp (hunk 3 above)
Comment (tridge) [reply libraries/GCS_MAVLink/GCS_Common.cpp]: done

**Review by peterbarker — APPROVED** [review]

File: libraries/AP_NavEKF3/AP_NavEKF3_PosVelFusion.cpp [hunk 4]
```diff
@@ -144,6 +144,42 @@ void NavEKF3_core::ResetPosition(resetDataSource posResetSource)
...
+
+    // set the variances using the position measurement noise parameter
+    P[7][7] = P[8][8] = sq(MAX(posAccuracy,frontend->_gpsHorizPosNoise));
+
+    // Correct the position for time delay relative to fusion time horizon assuming a constant velocity
+    // Limit time stamp to a range between current time and 5 seconds ago
+    const uint32_t timeStampConstrained_ms = MAX(MIN(timestamp_ms, imuSampleTime_ms), imuSampleTime_ms - 5000);
+    const int32_t delta_ms = int32_t(imuDataDelayed.time_ms - timeStampConstrained_ms);
+    const ftype delaySec = 1E-3F * ftype(delta_ms);
+    const Vector2F newPosNE = EKF_origin.get_distance_NE_ftype(loc) - stateStruct.velocity.xy() * delaySec;
```
Comment (RomanBapst) [inline libraries/AP_NavEKF3/AP_NavEKF3_PosVelFusion.cpp:175]: @priseborough @tridge I might look at this the wrong way but shouldn't this be a positive sign here?

**Pull request merged**